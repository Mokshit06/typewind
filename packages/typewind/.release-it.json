{
  "github": {
    "release": true
  },
  "git": {
    "commitMessage": "release: v${version}"
  }
}