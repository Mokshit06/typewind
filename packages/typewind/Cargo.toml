[package]
name = "typewind-swc"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["cdylib"]
path = "swc/lib.rs"

[profile.release]
lto = true

[dependencies]
serde = "1"
serde_json = "1.0.93"
swc_core = { version = "0.79.28", features = ["ecma_plugin_transform"] }
swc_ecma_codegen = "0.145.5"

# .cargo/config defines few alias to build plugin.
# cargo build-wasi generates wasm-wasi32 binary
# cargo build-wasm32 generates wasm32-unknown-unknown binary.
