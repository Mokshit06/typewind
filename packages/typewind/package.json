{
  "name": "typewind",
  "version": "0.1.3",
  "description": "The safety of Typescript with the magic of Tailwind.",
  "types": "dist/index.d.ts",
  "main": "dist/index.js",
  "homepage": "https://typewind.dev",
  "scripts": {
    "build": "npm run build:tsup && npm run build:cargo",
    "build:tsup": "tsup",
    "watch": "npm run build -- --watch",
    "release": "release-it",
    "build:cargo": "cargo build-wasi --release && cp target/wasm32-wasi/release/typewind_swc.wasm dist"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/mokshit06/typewind.git",
    "directory": "packages/typewind"
  },
  "bin": {
    "typewind": "dist/cli.js"
  },
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./dist/*": "./dist/*",
    "./babel": {
      "types": "./dist/babel.d.ts",
      "default": "./dist/babel.js"
    },
    "./transform": {
      "types": "./dist/transform.d.ts",
      "default": "./dist/transform.js"
    },
    "./swc": "./dist/typewind_swc.wasm"
  },
  "keywords": [
    "typescript",
    "tailwind",
    "tailwindcss",
    "babel-plugin",
    "zero-runtime"
  ],
  "author": "",
  "license": "MIT",
  "dependencies": {
    "@babel/core": "^7.20.12",
    "@babel/generator": "^7.20.14",
    "@babel/plugin-syntax-jsx": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "eval": "^0.1.8",
    "lightningcss": "^1.19.0",
    "tsup": "^6.5.0"
  },
  "devDependencies": {
    "@types/babel__core": "^7.20.0",
    "release-it": "^15.6.0"
  }
}
