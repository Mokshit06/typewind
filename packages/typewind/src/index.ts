declare const typewind: unique symbol;

// make brand/opaque type so that typescript lsp shows error as TypewindError<Message> and not Message
type TypewindError<T> = { [typewind]: T };

export declare const tw: TypewindError<"Typewind's types haven't been generated. Run `npx typewind generate` or follow the docs at https://typewind.dev/docs/installation">;
