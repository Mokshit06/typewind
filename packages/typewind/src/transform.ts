import * as babel from '@babel/core';
import { TransformerFn } from 'tailwindcss/types/config';

export const transformBabel = (ext: string, content: string) => {
  const config: babel.TransformOptions = {
    filename: `typewind.${ext}`,
    plugins: ['typewind/babel'],
  };
  if (ext === 'ts' || ext === 'tsx') {
    config.presets = ['@babel/preset-typescript'];
  }

  if (ext === 'js' || ext === 'jsx') {
    config.plugins?.push('@babel/plugin-syntax-jsx');
  }

  const res = babel.transformSync(content, config);

  if (res?.code) {
    return res?.code;
  }

  throw new Error('Failed to transform file');
}

export const typewindTransforms: Record<string, TransformerFn> = {
  tsx: (content) => transformBabel('tsx', content),
  ts: (content) => transformBabel('ts', content),
  jsx: (content) => transformBabel('jsx', content),
  js: (content) => transformBabel('js', content),
};
