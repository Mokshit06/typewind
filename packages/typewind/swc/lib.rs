use std::collections::HashMap;

use serde::Deserialize;
use swc_core::plugin::{plugin_transform, proxies::TransformPluginProgramMetadata};
use swc_core::{
    common::DUMMY_SP,
    ecma::{
        ast::Program,
        ast::*,
        transforms::testing::test,
        visit::{as_folder, FoldWith, VisitMut, VisitMutWith},
    },
};

#[derive(Deserialize)]
struct CandidateValueMap(HashMap<String, Vec<Vec<String>>>);

impl Default for CandidateValueMap {
    fn default() -> Self {
        setup_rule_map()
    }
}

fn fmt_to_tailwind(str: &str) -> String {
    str.chars()
        .map(|c| match c {
            '$' => '@',
            '_' => '-',
            c => c,
        })
        .collect()
}

fn inner_most_is_tw(expr: &Expr) -> bool {
    match expr {
        Expr::Ident(ident) => ident.sym == *"tw",
        Expr::Member(member_expr) => inner_most_is_tw(&*member_expr.obj),
        Expr::Call(call_expr) => call_expr
            .callee
            .as_expr()
            .and_then(|expr| Some(inner_most_is_tw(&**expr)))
            .unwrap_or(false),
        _ => return false,
    }
}

fn analyse_expr(v: &mut TransformVisitor, wr: &mut Vec<String>, e: &Expr) {
    match e {
        Expr::Member(member_expr) => {
            let style = match &member_expr.prop {
                MemberProp::Ident(ident) => fmt_to_tailwind(ident.sym.as_ref()),
                MemberProp::Computed(computed) => {
                    match computed.expr.as_lit().expect("Only literal values allowed") {
                        Lit::Str(str) => str.value.to_string(),
                        _ => panic!("Only string literal values allowed"),
                    }
                }
                _ => panic!("Private values not allowed"),
            };

            if member_expr.prop.is_computed() {
                v.style_queue.push_str(&style);
            } else if !v.style_queue.is_empty() {
                let maybe_arbitrary_val = v.style_queue.as_str();
                let without_suffix = &style[..style.len() - 1];

                let CandidateValueMap(rule_map) = &v.rule_map;

                // @ means $ here as for opacity or named container queries etc.
                // its being replaced to @ by the fmt_to_tailwind fn
                if style.ends_with("@") {
                    wr.push(format!(
                        "{}{}/{}",
                        v.get_prefix(),
                        without_suffix,
                        maybe_arbitrary_val
                    ));
                } else if let Some(rules) = rule_map.get(without_suffix) {
                    let is_known_value = rules.iter().any(|rule| {
                        if rule.contains(&maybe_arbitrary_val.to_string()) {
                            true
                        } else {
                            false
                        }
                    });

                    if is_known_value {
                        wr.push(format!(
                            "{}{}{}",
                            v.get_prefix(),
                            style,
                            maybe_arbitrary_val
                        ));
                    } else {
                        wr.push(format!(
                            "{}{}[{}]",
                            v.get_prefix(),
                            style,
                            maybe_arbitrary_val
                        ));
                    }
                }

                v.style_queue.clear()
            } else {
                wr.push(format!("{}{}", v.get_prefix(), style));
            }

            analyse_expr(v, wr, &*member_expr.obj)
        }
        Expr::Call(call_expr) => {
            let CallExpr { callee, args, .. } = call_expr;

            let callee_member = callee
                .as_expr()
                .expect("callee can only be expr")
                .as_member()
                .expect("only member expr allowed");

            analyse_expr(v, wr, &callee_member.obj);

            let style = fmt_to_tailwind(
                callee_member
                    .prop
                    .as_ident()
                    .expect("callee can only be identifier")
                    .sym
                    .as_ref(),
            );

            if style == "variant" {
                let prefix = match &args
                    .get(0)
                    .expect("variant not specified")
                    .expr
                    .as_lit()
                    .expect("Only literal values allowed")
                {
                    Lit::Str(str) => format!("[{}]:", str.value.as_ref()),
                    _ => panic!("Only string literal values allowed"),
                };

                v.cur_prefix.push(prefix);

                analyse_expr(v, wr, &*args.get(1).expect("style not specified").expr);

                v.cur_prefix.pop();
            } else {
                v.cur_prefix.push(if style == "important" {
                    "!".to_string()
                } else {
                    format!("{}:", style)
                });

                analyse_expr(v, wr, &*args.get(0).expect("style not specified").expr);

                v.cur_prefix.pop();
            }
        }
        _ => {}
    }
}

#[derive(Default)]
pub struct TransformVisitor {
    cur_prefix: Vec<String>,
    inside_tw: bool,
    style_queue: String,
    rule_map: CandidateValueMap,
}

impl TransformVisitor {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn get_prefix(&self) -> String {
        self.cur_prefix.join("")
    }
}

impl VisitMut for TransformVisitor {
    fn visit_mut_expr(&mut self, e: &mut Expr) {
        if (e.is_call() || e.is_member()) && !self.inside_tw && inner_most_is_tw(&e) {
            self.inside_tw = true;

            let mut wr = Vec::<String>::new();
            analyse_expr(self, &mut wr, &e);

            *e = Expr::Lit(Lit::Str(Str {
                span: DUMMY_SP,
                value: wr.join(" ").into(),
                raw: None,
            }));

            self.inside_tw = false;
        } else {
            e.visit_mut_children_with(self)
        }
    }
}

#[plugin_transform]
pub fn process_transform(program: Program, _metadata: TransformPluginProgramMetadata) -> Program {
    program.fold_with(&mut as_folder(TransformVisitor::new()))
}

fn setup_rule_map() -> CandidateValueMap {
    serde_json::from_str::<CandidateValueMap>(include_str!("./value.json")).unwrap()
}

test!(
    Default::default(),
    |_| as_folder(TransformVisitor::new()),
    boo,
    // Input codes
    r#"let style = tw.flex.$lg(tw.bg_black$['20']).md(tw.important(tw.works).text_["18px"].text_["red-200"])"#,
    // Output codes after transformed with plugin
    r#"let style = "flex @lg:bg-black/20 md:text-red-200 md:text-[18px] md:!works""#
);
