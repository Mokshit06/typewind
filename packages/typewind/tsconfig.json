{
  "include": [
    "src/**/*.ts"
  ],
  "exclude": [
    "node_modules",
    "dist"
  ],
  "compilerOptions": {
    "isolatedModules": true,
    "esModuleInterop": true,
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "target": "ESNext",
    "strict": true,
    "allowJs": true,
    "skipLibCheck": true
  }
}