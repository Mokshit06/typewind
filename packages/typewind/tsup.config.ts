import { defineConfig } from 'tsup';
import { createRequire } from 'module';

export default defineConfig({
  entry: [
    'src/index.ts',
    'src/cli.ts',
    'src/babel.ts',
    'src/evaluate.ts',
    'src/transform.ts',
    'src/cn.ts',
  ],
  splitting: false,
  clean: true,
  platform: 'node',
  external: ['typewind', 'tailwindcss'],
  format: ['cjs', 'esm'],
  target: 'esnext',
  dts: true,
});
