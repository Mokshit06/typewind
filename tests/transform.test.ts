// @ts-ignore
import { transformBabel } from 'typewind/transform';
import * as glob from 'glob';
import * as path from 'path';
import * as fs from 'fs';

function setupTest(fixturesDir: string, dir: string) {
  const dirPath = path.join(fixturesDir, dir);
  process.chdir(dirPath);

  const files = glob.sync('*.{ts,tsx,js,jsx}', { cwd: dirPath });

  files.forEach(file => {
    const filePath = path.join(dirPath, file);
    const ext = path.extname(file).slice(1);
    const content = fs.readFileSync(filePath, 'utf-8');
    const res = transformBabel(ext, content);
    expect(res).toBeDefined();
    expect(res).toMatchSnapshot();
  });
}

describe('transformBabel', () => {
  const cwd = process.cwd();
  const fixturesDir = path.join(cwd, 'tests', 'fixtures');

  test('works with container-queries', () => {
    setupTest(fixturesDir, 'container-queries');
  });

  test('works with arbitrary-values', () => {
    setupTest(fixturesDir, 'arbitrary-values');
  });

  test('works with arbitrary-variants', () => {
    setupTest(fixturesDir, 'arbitrary-variants');
  });

  test('works with modifiers', () => {
    setupTest(fixturesDir, 'modifiers');
  });

  test('works with normal-usage', () => {
    setupTest(fixturesDir, 'normal-usage');
  });

  test('works with transform', () => {
    setupTest(fixturesDir, 'transform');
  });
});
