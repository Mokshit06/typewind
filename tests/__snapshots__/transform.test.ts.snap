// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformBabel works with arbitrary-values 1`] = `
"import { tw } from 'typewind';
export default function Button() {
  return <button className={"text-[20px]"}>Click Me</button>;
}"
`;

exports[`transformBabel works with arbitrary-variants 1`] = `
"import { tw } from 'typewind';
export default function Button() {
  return <button className={"[&:nth-child(3)]:underline"}>Click Me</button>;
}"
`;

exports[`transformBabel works with container-queries 1`] = `
"import { tw } from 'typewind';
export default function App() {
  return <div className={"@container"}>
      <div className={"@lg:underline"}>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Animi, non.
      </div>
    </div>;
}"
`;

exports[`transformBabel works with modifiers 1`] = `
"import { tw } from 'typewind';
export default function Button() {
  return <button className={"bg-blue-500 hover:bg-blue-600 text-white rounded py-3 px-4 md:py-4 md:px-5 lg:py-5 lg:px-6 dark:bg-sky-900 dark:hover:bg-sky-800 !text-red-500 hover:!text-red-500"}>
      Click Me
    </button>;
}"
`;

exports[`transformBabel works with normal-usage 1`] = `
"import { tw } from 'typewind';
import './index.css';
export default function App() {
  return <div className={"h-screen w-full bg-gray-800 flex justify-center items-center"}>
      <p className={"text-white"}>Hello World!</p>
    </div>;
}"
`;

exports[`transformBabel works with transform 1`] = `
"import { tw } from 'typewind';
export default function Button() {
  console.log("text-red-100");
  return <button className={"text-red-100"}>Click Me</button>;
}"
`;
